gateway/src/service.rs

```diff
@@ -0,0 +1,522 @@
+
+use crate::auth::User;
+use crate::{auth::Key, AccountName};
+
+impl GatewayService {
+    /// Initialize `GatewayService` and its required dependencies.
+    ///
+    /// * `args` - The [`Args`] with which the service was
+    /// started. Will be passed as [`Context`] to workers and state.
+    pub async fn init(args: Args) -> Arc<Self> {
+        info!("docker host: tcp://127.0.0.1:2735");
+        let docker = Docker::connect_with_http_defaults().unwrap();
+
+        let hyper = HyperClient::new();
+        if !StdPath::new(DB_PATH).exists() {
+            Sqlite::create_database(DB_PATH).await.unwrap();
+        }
+
+        info!(
+            "state db: {}",
+            std::fs::canonicalize(DB_PATH).unwrap().to_string_lossy()
+        );
+        let db = SqlitePool::connect(DB_PATH).await.unwrap();
+
+        query("CREATE TABLE IF NOT EXISTS projects (project_name TEXT PRIMARY KEY, account_name TEXT NOT NULL, initial_key TEXT NOT NULL, project_state JSON NOT NULL)")
+            .execute(&db)
+            .await
+            .unwrap();
+
+        query("CREATE TABLE IF NOT EXISTS accounts (account_name TEXT PRIMARY KEY, key TEXT UNIQUE, super_user BOOLEAN DEFAULT FALSE)")
+            .execute(&db)
+            .await
+            .unwrap();
+
+        let service = Arc::new(Self {
+            docker,
+            hyper,
+            db,
+            sender: Mutex::new(None),
+            args,
+        });
+
+        let worker = Worker::new(Arc::clone(&service));
+        let sender = worker.sender();
+        service.set_sender(Some(sender)).await;
+        tokio::spawn({
+            let service = Arc::clone(&service);
+            async move {
+                match worker.start().await {
+                    Ok(_) => info!("worker terminated successfully"),
+                    Err(err) => error!("worker error: {}", err),
+                };
+                service.set_sender(None).await;
+            }
+        });
+
+        // Queue up all the projects for reconciliation
+        for Work {
+            project_name,
+            account_name,
+            work,
+        } in service
+            .iter_projects()
+            .await
+            .expect("could not list projects")
+        {
+            match work.refresh(&service.context()).await {
+                Ok(work) => service
+                    .send(Work {
+                        project_name,
+                        work,
+                        account_name,
+                    })
+                    .await
+                    .expect("failed to queue work at startup"),
+                Err(err) => error!("could not refresh state for user=`{account_name}` project=`{project_name}`: {}. Skipping it for now.", err)
+            }
+        }
+
+        service
+    }
+
+    pub async fn set_sender(&self, sender: Option<Sender<Work>>) -> Result<(), Error> {
+        *self.sender.lock().await = sender;
+        Ok(())
+    }
```

Comment:
Does this need to return a `Result`?
```suggestion
    pub async fn set_sender(&self, sender: Option<Sender<Work>>) {
        *self.sender.lock().await = sender;
    }
```